Show template center UI when no block is selected

jeryj opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/56194


## What?
<!-- In a few words, what is the PR actually doing? -->

When using the Template Editor from within a post, the central template ui to go back to the post was not showing after selecting a block in the canvas. We should only be collapsing this area when we're:
- editing a template
- top toolbar mode is on
- a block is selected
- a user has clicked to collapse the area
- the viewport is large

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Fixes a bug where the central template ui was hidden.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Check for the correct variables before hiding the UI:
- editing a template
- top toolbar mode is on
- a block is selected
- a user has clicked to collapse the area
- the viewport is large

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
(Partially stealing these from @afercia's issue - thanks for the clear screenshots!)
- Edit a post.
- In the setting panel, click the current template. A popover open. Click 'Edit template'. Screenshot:

![Screenshot 2023-11-16 at 10 39 48](https://github.com/WordPress/gutenberg/assets/1682452/99b3d631-a3f7-4a9d-a3eb-9962730a0f0e)

- The editor switches to a 'Edit template' mode.
- Observe in the top bar there is a 'Back' button to switch back to the normal Post editor mode.
- Observe that whatever you do, the 'Back' button stays there.
- Switch to Top Toolbar mode
- Central area should remain (unless a block is selected)
- Unselecting a block should show the central area
- Selecting a block should show the top toolbar and the central area should be hidden
- Click the Collapse button when a block is selected
- Block tools should collapse, and the central area should show again to be able to go back to the post
- Try to break it :) 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jeryj committed:
Show template center ui when no block is selected on top toolbar on template post editor

The central template ui to go back to the post was not showing after selecting a block in the canvas. We should only be collapsing this area when we're:
- editing a template
- top toolbar mode is on
- a block is selected
- a user has clicked to collapse the area
- the viewport is large

afercia commented:
@jeryj thanks for looking into this!

Regarding this part:
```
Switch to Top Toolbar mode
Central area should remain (unless a block is selected)
Unselecting a block should show the central area
```

How do you unselect a block? I'm clicking in the empty space after the footer but for me the central area with the Back button stays hidden, even if no block is selected. Screenshot:

![Screenshot 2023-11-20 at 12 52 10](https://github.com/WordPress/gutenberg/assets/1682452/fdc523a4-65c0-4025-8994-05a3af84e553)

tellthemachines reviewed (commented):
Thanks for fixing this! The center is now always visible when _not_ in top toolbar mode, but as @afercia noticed it still doesn't show when in top toolbar mode and no block is selected. Adding a check to see if a block is selected around line 155 could perhaps fix that?

Another thing I noticed is the tooltip for "Show block tools" is showing when text labels are enabled, though that can probably be fixed separately:
<img width="781" alt="Screenshot 2023-12-04 at 1 49 01 pm" src="https://github.com/WordPress/gutenberg/assets/8096000/c533a1c6-406e-458d-84e3-702d2f719d04">

Also the tiny niggling detail of the focus ring for "Show block tools" being partially hidden by the center area 😅

afercia commented:
> Another thing I noticed is the tooltip for "Show block tools" is showing when text labels are enabled, t

Thanks for noticing this. I'd think this should really be automated in the component, as in: when there is visible text and the same text is going to be used for the tooltip, don't show the tooltip. I suggested this behavior to @ciampo but we didn't came to an agreement. The point is this keeps to be missed because we often forget to test the 'show text labels' preference. Automating the behavior would certainly help.

ciampo commented:
`Button` shouldn't be aware of the "show text labels" options.

Comparing `Button`'s contents to the tooltip's contents is also not trivial, since `Button`'s contents can be any DOM sub-tree, and I can foresee many edge cases where it wouldn't be straightforward to compare them to the tooltip — for example:

- it's not easy to determine what's visible and what's not visible
- it's not easy to determine what contributes to the accessible name of the button
- there could be slight space / casing variations
- etc

Also, if there was ever a component where this feature should be implemented, I wonder if `Tooltip` itself would be a better fit.

tellthemachines commented:
> `Button` shouldn't be aware of the "show text labels" options.

Independently of which specific components are aware of the feature, I do think this functionality should be dealt with in the components library, because it's an essential part of ensuring accessibility, and it needs to behave consistently across the whole interface. 

The current implementation of the text labels preference depends on a classname output on the body element, which means Button component doesn't change in itself when its label changes from icon to text: the change is done via CSS only. Adjustments such as removing the tooltip need to be passed to Button conditionally depending on the preference setting, but that requires every single component that renders a Button to know about the preference, and every single developer using a Button to be aware of it. 

It's the last part that is most concerning, because it requires extensive education and increases the overhead for core contributors and extenders alike. The point of a components library is to provide tools that decrease the mental load and decision-making of developers, and make it easier to build a consistent interface. 

The other advantage to making the components library responsible for handling the button text labels preference is that component developers will have to take this into account for all new components, avoiding situations where the design of components makes it impossible to implement accessible labels after the fact. See #46299 for an idea of the difficulties with implementing text labels for the block inspector tools.

The main obstacle to this is that preferences are editor-specific, and components shouldn't be aware of the environment they're rendering in. One option would be to try a similar approach to the Interface package, where the [preferences store is called](https://github.com/WordPress/gutenberg/blob/trunk/packages/interface/src/store/selectors.js#L19) with a scope that has to be given by the component rendering it. But having to pass in a scope for each component used isn't practical; components are often used by other generic packages such as block-editor.

It would be nice if there were a way of detecting what scope/preferences apply in the current context and have components render text/icons depending on the preference set. I'm unsure if there's been any thinking around this already? Cc @talldan who worked on preferences quite a bit.

I imagine a similar mechanism would be necessary for implementing dark mode in components too, if that is open to consideration (dark mode across the WP interface would be much welcome accessibility enhancement!)

afercia commented:
I'd second what @tellthemachines said, as in the components library should be responsible, in some way, for handling the button text labels preference.

> `Button` shouldn't be aware of the "show text labels" options.

@ciampo I do understand your arguments. But, the point is there have been several cases where specific implementations of Button didn't take into account the 'show text labels' preference. This keeps to be missed by developers and broke several times. I'd think we need a pragmatical solution to an actual problem.

jeryj committed:
Add hasBlockSelected check to edit post template editor

jeryj commented:
> Thanks for fixing this! The center is now always visible when not in top toolbar mode, but as @afercia noticed it still doesn't show when in top toolbar mode and no block is selected. Adding a check to see if a block is selected around line 155 could perhaps fix that?

Added the `hasBlockSelected` check, and it does fix the issue. Thanks! I had added that same fix [in the site editor](https://github.com/WordPress/gutenberg/pull/56219) but forgot to add it here.

tellthemachines reviewed (approved):
Thanks for the update, this is working very well now!

ciampo commented:
Thank you @tellthemachines for sharing your perspective!

> The main obstacle to this is that preferences are editor-specific, and components shouldn't be aware of the environment they're rendering in.

As you also mentioned, this is the main problem. The components package shouldn't be aware of any editor preferences.

> Adjustments such as removing the tooltip need to be passed to Button conditionally depending on the preference setting, but that requires every single component that renders a Button to know about the preference, and every single developer using a Button to be aware of it.

I also agree that this is not great devEx and can lead to poor results.

> I imagine a similar mechanism would be necessary for implementing dark mode in components too, if that is open to consideration (dark mode across the WP interface would be much welcome accessibility enhancement!)

We've been actually working on this, and the solution that we're landing on is a theme provider. We initially experimented with a `Theme` component internal to the component package, but the scope of it is large enough that we're now considering creating a separate [`@wordpress/theme` package](https://github.com/WordPress/gutenberg/pull/56669).

So, maybe, the theme could be the means through which we can expose the editor's preference to the components in the package.

@afercia 

>  when there is visible text and the same text is going to be used for the tooltip, don't show the tooltip.

Regarding this specific requirement, I don't think that we can reliably find a way to implement this logic. Implementing such a check is not trivial, prone to failing over many edge cases, a cause for inflated bundle size and extra runtime costs, similarly to what discussed in https://github.com/WordPress/gutenberg/discussions/56231#discussioncomment-7801125

afercia commented:
> Regarding this specific requirement, I don't think that we can reliably find a way to implement this logic. Implementing such a check is not trivial, prone to failing over many edge cases, a cause for inflated bundle size and extra runtime costs, similarly to what discussed in [#56231 (reply in thread)](https://github.com/WordPress/gutenberg/discussions/56231#discussioncomment-7801125)

@ciampo thanks for your feedback. I do understand your concerns but I'd kindly note that you are not providing an alternative solution. We do have a problem. We need to solve it. The 'show text labels' preference broke several times in the past and that's going to happen again, as apparently contributors to this project aren't well aware of it and aren't educated to check for it. Saying no to a specific proposed approach is perfectly fine but blocking the solution of a problem isn't ideal. How can we solve this problem? Thanks.

ciampo commented:
> I'd kindly note that you are not providing an alternative solution
> [...]
> Saying no to a specific proposed approach is perfectly fine but blocking the solution of a problem isn't ideal.

I am highlighting a potential high-level way forward for how to "connect" components to the "show text label" preferences, inspired by @tellthemachines 's suggestion :

> > I imagine a similar mechanism would be necessary for implementing dark mode in components too, if that is open to consideration (dark mode across the WP interface would be much welcome accessibility enhancement!)
> 
> We've been actually working on this, and the solution that we're landing on is a theme provider. We initially experimented with a `Theme` component internal to the component package, but the scope of it is large enough that we're now considering creating a separate [`@wordpress/theme` package](https://github.com/WordPress/gutenberg/pull/56669).
> 
> So, maybe, the theme could be the means through which we can expose the editor's preference to the components in the package.

What I _don't_ have a (good) solution for is the request to automatically detect when the contents of a tooltip popup match the contents of the tooltip trigger, which could have been a more generic, component-centric approach.
